src/vtok_agent/src/imds.rs

```diff
@@ -124,20 +124,41 @@ impl ImdsCache {
     }
 
     fn fetch_role_arn() -> Result<String, Error> {
-        let profile_arn =
-            serde_json::from_str::<IamInfo>(&Self::fetch("/latest/meta-data/iam/info")?)
-                .and_then(|info| Ok(info.instance_profile_arn))
-                .map_err(Error::ParseError)?;
         let role_name = Self::fetch_role_name()?;
 
-        let role_arn = profile_arn
-            .split(":")
-            .map(|word| match word.find("instance-profile/") {
-                Some(_) => format!("role/{}", role_name),
-                None => word.to_string(),
-            })
-            .collect::<Vec<String>>()
-            .join(":");
+        let output = Command::new("aws")
+            .arg("iam")
+            .arg("get-role")
+            .arg("--role-name")
+            .arg(format!("{}", role_name))
+            .output()
+            .map_err(Error::IoError)?;
+        // If error, fallback to the previous way of fetching the IAM role arn.
+        // Note: This logic would not work for IAM roles which include paths.
+        if !output.status.success() {
+            let profile_arn =
+                serde_json::from_str::<IamInfo>(&Self::fetch("/latest/meta-data/iam/info")?)
+                    .and_then(|info| Ok(info.instance_profile_arn))
+                    .map_err(Error::ParseError)?;
+
+            let role_arn = profile_arn
+                .split(":")
+                .map(|word| match word.find("instance-profile/") {
+                    Some(_) => format!("role/{}", role_name),
+                    None => word.to_string(),
+                })
+                .collect::<Vec<String>>()
+                .join(":");
+
+            return Ok(role_arn);
+        }
+
+        let iam_role_json: serde_json::Value =
+            serde_json::from_str(&String::from_utf8(output.stdout).map_err(Error::Utf8Error)?)
+                .map_err(Error::ParseError)?;
+
+        let role_arn = iam_role_json["Role"]["Arn"].as_str().unwrap().to_string();
```

Comment:
Should comment on why is it safe to `unwrap` here. i.e.

Reply:
I updated the patch to include a custom data structure instead and parse the json output using that struct format.

Thanks for review.